Determine electron production rate using the Fang et al models.

ykempf opened the pull request:
I'll put this up against my field_tracing branch until we settle on [#fmihpc/625](https://github.com/fmihpc/vlasiator/pull/625).

ykempf commented on common.h:
```diff
@@ -491,6 +493,7 @@ namespace physicalconstants {
    const Real MASS_ELECTRON = 9.10938188e-31; /**< Electron rest mass, units: kg.*/
    const Real MASS_PROTON = 1.67262158e-27; /*!< Proton rest mass, units: kg.*/
    const Real R_E = 6.3712e6; /*!< radius of the Earth, units: m. */
+   const Real g = 9.80665; /*!< Standard acceleration of gravity at Earth's surface, units: m s^-2.*/
```
Taken from WIkipedia I think – the digits here will never be critical in this code...

ursg replied:
My fist thought when I saw this was that it would be the quantum mechanical [g-factor](https://en.wikipedia.org/wiki/G-factor_(physics)) instead of Earth's surface gravity. Should we rename this to "EarthGravity", just to avoid confusion?

ykempf commented on sysboundary/ionosphere.h:
```diff
@@ -148,20 +151,130 @@ namespace SBC {
+      inline Real fangNormalizedColumnMass(cint p, creal E, cuint h) {
+         Real y;
+         switch(p) {
+            case Particles::PROTON:
+               y = 7.5 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 1e-4 / 10, 0.9);
+               break;
+            case Particles::ELECTRON:
+               y = 2 / E * pow(atmosphere[h].density * atmosphere[h].scaleHeight / 6e-6 / 10, 0.7);
+               break;
+            default:
+               cerr << "Invalid precipitating species!" << endl;
+               abort();
+         }
+         return y;
+      }
+      
+      /*!< Energy dissipation f as a function of altitude and energy E, for
+       * - electrons Fang 2010 https://doi.org/10.1029/2010GL045406 equation (4)
+       * - protons Fang 2013 https://doi.org/10.1002/jgra.50484 equation (6)
+       * 
+       * Note that the i indices are taken -1 as we are 0-indexing in the arrays P_ij.
+       * 
+       * \param p particle species enum Particles::PROTON and ELECTRON
+       * \param E energy in keV
+       * \param h altitude index
+       */
+      inline Real fangEnergyDissipation(cint p, creal E, cint h) {
+         creal y = fangNormalizedColumnMass(p, E, h);
+         
+         cint offset = p*12; // 0 for proton, 12 for electron, see indexing in P_ij
+         return
+           fC(offset+0,E) * pow(y, fC(offset+1,E)) * exp(-fC(offset+ 2,E) * pow(y, fC(offset+ 3,E)))
+         + fC(offset+4,E) * pow(y, fC(offset+5,E)) * exp(-fC(offset+ 6,E) * pow(y, fC(offset+ 7,E)))
+         + fC(offset+8,E) * pow(y, fC(offset+9,E)) * exp(-fC(offset+10,E) * pow(y, fC(offset+11,E))); // this line returns zero for electrons, see P_ij
+      }
+      
+      Real totalIonizationFang(
+         cint n,
+         cint h
+      );
```
I put this chain of function implementing the models in the header as inline, if this should be done in another way architecturally, let me know.

ursg replied:
We are starting to collect so many ionization models now that we might even stat spinning *them* off into separate source files, too.

ursg replied:
(That way, they could fathomably also be simply included in the sigmaProfiles miniapp, without the need for any code duplication anymore)

ykempf commented on vlasiator.cpp:
```diff
@@ -1023,7 +1023,8 @@ int main(int argn,char* args[]) {
          SBC::ionosphereGrid.mapDownBoundaryData(perBGrid, dPerBGrid, momentsGrid, volGrid, technicalGrid);
-         SBC::ionosphereGrid.calculateConductivityTensor(SBC::Ionosphere::F10_7, SBC::Ionosphere::recombAlpha, SBC::Ionosphere::backgroundIonisation);
+         SBC::ionosphereGrid.calculateProtonPrecipitation(mpiGrid);
+         SBC::ionosphereGrid.calculateConductivityTensor(SBC::Ionosphere::F10_7, SBC::Ionosphere::recombAlpha, SBC::Ionosphere::backgroundIonization);
```
This is not directly related but it's useless to pass down the SBC::ionosphereGrid's own parameters as arguments to this function. If you wish I can remove them and just call them within the class code.

ykempf commented on sysboundary/ionosphere.cpp:
```diff
@@ -851,13 +879,288 @@ namespace SBC {
+                     const Real energy = 0.5 * mass * normV*normV / (1000 * physicalconstants::CHARGE); // in keV, as the limits of the bins
+                     
+                     // Find the correct energy bin number to update
+                     int binNumber = round((log(energy) - log(productionMinProtonEnergy)) / log(productionMaxProtonEnergy/productionMinProtonEnergy) * (productionNumProtonEnergies-1));
+                     binNumber = max(binNumber, 0); // anything < emin goes to the lowest channel
+                     binNumber = min(binNumber, productionNumProtonEnergies-1); // anything > emax goes to the highest channel
+                     
+                     // TODO check the bin number or -1
```
So I think that on 1022 it should not have a minus 1 (that is copied from Maxime's code where the indexing, I believe, is handled slightly differently – possibly incorrectly there imho).

ykempf commented on sysboundary/ionosphere.cpp:
```diff
@@ -851,13 +879,288 @@ namespace SBC {
+            }
+            
+            // Averaging within each bin and conversion to unit of part. cm^-2 s^-1 sr^-1 keV^-1
+            // What TODO with the sr?
```
So I'd like expert opinions here, I didn't wrap my head around what to make of this `sr^-1` here and possibly I thus handle the solid angles/loss cone stuff incorrectly. Please enlighten me.

ykempf commented on sysboundary/ionosphere.h:
```diff
@@ -148,20 +151,130 @@ namespace SBC {
+      );
       
       // Hardcoded constants for calculating ion production table
       // TODO: Make these parameters?
```
Just a note here, to avoid mixing/breaking things I kept the lookup table parameters from the Rees/ Sergienko & Ivanov paths separate from the Fang one so far. We can unify and indeed then make them all cfg parameters later. Actually I think that the energy ranges shouldn't be touched for the Fang models as they are now by default the stated validity range of the parametrisations.

ursg replied:
I guess it would simplify the code a lot if the other models, too, were to get rid of the precomputed lookup table.
We would need to measure the performance impact of that though (I'd wager it is not large)

ursg commented on sysboundary/ionosphere.cpp:
```diff
@@ -75,7 +75,7 @@ namespace SBC {
    Real Ionosphere::couplingInterval; // Ionosphere update interval
    int Ionosphere::solveCount; // Counter of the number of solvings
-   Real Ionosphere::backgroundIonisation; // Background ionisation due to stellar UV and cosmic rays
+   Real Ionosphere::backgroundIonization; // Background ionization due to stellar UV and cosmic rays
```
Hey, I liked my british spelling! :D

ykempf replied:
We used to have the policy, inherited from Ilja, to use American spelling cuz it's less typin. Feel free to overhaul that but it can cause headaches later...

ursg replied:
Nah, I'm fully ok with that y'all.

ursg commented on sysboundary/ionosphere.cpp:
```diff
@@ -629,7 +629,7 @@ namespace SBC {
       Real prevDensity = 0;
       Real prevAltitude = 0;
-      std::vector<std::array<Real, 5>> MSISvalues;
+      std::vector<std::array<Real, 6>> MSISvalues;
```
This would also be a prime candidate to have an enum instead of a hardcoded number.
Or how about a struct, even!

ykempf replied:
Yep. Here btw we had the discussion on which levels to use and why not all, but that branch/commit of mine disappeared into some unhola. What's your opinion on that?

ursg commented on sysboundary/ionosphere.cpp:
```diff
@@ -684,20 +690,25 @@ namespace SBC {
          atmosphere[h].parallelcoeff = sigma_e;
       }
 
+      // If we don't use the Rees or Sergienko & Ivanov parametrisations, we don't need the lookup table
```
I mean... technically... we don't even need that in the other cases, either. I just adopted the same lookup table as GUMICS had, because I assumed there's some point to it. There probably was on the supercomputers of the mid 90'ies, but we could just as well rewrite the other models to simply be evaluated on-the-spot.
Might not even be slower, in the end, and would have taken a lot less debugging.

ykempf replied:
Uh. Argh. And I guess you'd like *me* to tackle that?

ursg replied:
Let's leave it like this, for now. The PR is big enough already.

ursg commented on sysboundary/ionosphere.cpp:
```diff
@@ -851,13 +879,288 @@ namespace SBC {
+                     
+                     const Real normV = sqrt(VX*VX + VY*VY + VZ*VZ);
+                     const Real VdotB_norm = (B[0]*VX + B[1]*VY + B[2]*VZ)/normV;
+                     Real countAndGate = floor(VdotB_norm/cosAngle);  // gate function: 0 outside loss cone, 1 inside
```
I guess here we could decide to either use a smoother gate function (a smoothstep with a cell-width transition maybe?), so that cells at the edge of the loss cone are still partially counted.
Or then we could do a proper cube-cone intersection test. I'm sure computer graphics has some efficient algorithms up the sleeve for that already, since it is a key ingredient in voxel cone tracing for indirect lighting in computer games... :)

ykempf replied:
That would be cool if you had, I namely had the issue of no precipitation if dv isn't small enough...

ursg commented on sysboundary/ionosphere.h:
```diff
@@ -148,20 +151,130 @@ namespace SBC {
+       *   Lines 12-19 Fang 2010 electron Pij coefficients https://doi.org/10.1029/2010GL045406
+       *   Lines 20-23 Padding with zeros to simplify calls.
+       */
+      constexpr static std::array<std::array<Real, 4>, 24> P_ij = {{ // hooray for C++ initialization, the additional {} is needed.
+         // proton
+         { 2.55050E+0,   2.69476E-1,  -2.58425E-1,   4.43190E-2},
+         { 6.39287E-1,  -1.85817E-1,  -3.15636E-2,   1.01370E-2},
+         { 1.63996E+0,   2.43580E-1,   4.29873E-2,   3.77803E-2},
+         {-2.13479E-1,   1.42464E-1,   1.55840E-2,   1.97407E-3},
+         {-1.65764E-1,   3.39654E-1,  -9.87971E-3,   4.02411E-3},
+         {-3.59358E-2,   2.50330E-2,  -3.29365E-2,   5.08057E-3},
+         {-6.26528E-1,   1.46865E+0,   2.51853E-1,  -4.57132E-2},
+         { 1.01384E+0,   5.94301E-2,  -3.27839E-2,   3.42688E-3},
+         {-1.29454E-6,  -1.43623E-1,   2.82583E-1,   8.29809E-2},
+         {-1.18622E-1,   1.79191E-1,   6.49171E-2,  -3.99715E-3},
+         { 2.94890E+0,  -5.75821E-1,   2.48563E-2,   8.31078E-2},
+         {-1.89515E-1,   3.53452E-2,   7.77964E-2,  -4.06034E-3},
+         // electron
+         { 1.24616E+0,   1.45903E+0,  -2.42269E-1,   5.95459E-2},
+         { 2.23976E+0,  -4.22918E-7,   1.36458E-2,   2.53332E-3},
+         { 1.41754E+0,   1.44597E-1,   1.70433E-2,   6.39717E-4},
+         { 2.48775E-1,  -1.50890E-1,   6.30894E-9,   1.23707E-3},
+         {-4.65119E-1,  -1.05081E-1,  -8.95701E-2,   1.22450E-2},
+         { 3.86019E-1,   1.75430E-3,  -7.42960E-4,   4.60881E-4},
+         {-6.45454E-1,   8.49555E-4,  -4.28581E-2,  -2.99302E-3},
+         { 9.48930E-1,   1.97385E-1,  -2.50660E-3,  -2.06938E-3},
+         {          0,            0,            0,            0},
+         {          0,            0,            0,            0},
+         {          0,            0,            0,            0},
+         {          0,            0,            0,            0}
+      }};
+      
```
Now this is really ugly.
Why does this need to be a single array, in which electrons just *happen* to start at index 13?
Can't this be a 2-dimensional array, with the just previously defined enum as an index? Or two separate ones altogether?

ursg replied:
Also, I suggest this be called fang_P_ij

ykempf replied:
Uh, sure, this has evolved organically while I was writing out the papers in code form.

ursg commented on sysboundary/ionosphere.h:
```diff
@@ -148,20 +151,130 @@ namespace SBC {
       constexpr static int productionNumAccEnergies = 60;
       constexpr static int productionNumTemperatures = 60;
-      constexpr static int productionNumParticleEnergies = 100;
+// moved to be a parameter below       constexpr static int productionNumElectronEnergies = 100;
```
Just delete this lnie.

ykempf replied:
Jawohl.

ursg commented on sysboundary/ionosphere.h:
```diff
@@ -148,20 +151,130 @@ namespace SBC {
       constexpr static Real ion_electron_T_ratio = 4.; // TODO: Make this a parameter (and/or find value from kinetics)
-      // Ionoisation production table
+      
+      // These are parameters already
```
In 5 years, somebody will read this comment and be very confused by it, and it doesn't help code understanding at all.. so I suggest to just drop it.

ykempf replied:
Wilco.

ursg reviewed (changes requested):
Even though I still haven't come around to thoroughly checking your maths (especial with the steradian factors), I'm dispatching this review already now. Because otherwise, it will just stay pending in my open browser window for another two weeks or so.

Anyway: Nice. A comparison plot of the three models' electron densities (height resolved) would be interesting.

lkotipal committed:
Use epsilon to get rid of infinities

lkotipal committed:
Clip alpha_1 instead of scaling constituents

lkotipal committed:
Output amount of cells refined/unrefined

lkotipal committed:
Whoops

lkotipal committed:
Refinement param for minimum portion of grid to refine

lkotipal committed:
Also log actual refine amount

lkotipal committed:
Only count cells to refine

lkotipal committed:
Use total B for calculating alpha

ykempf committed:
Make in-code and DRO derivatives be actual derivatives and not differences.

ykempf committed:
Fix a Real -> TReal conversion in fieldtracing.

ykempf committed:
Exclusive node for carrington CI testpackage script.

ykempf committed:
Two forgotten multiplications by DX in reconstruction coefficients.

lkotipal committed:
Use char literals for diff

lkotipal committed:
Less stupid regex

ursg committed:
Merge pull request #729 from lkotipal/fix-version

Fix generate_version

ursg committed:
Greatly simplify Makefile

the Makefile now makes use of gcc's automatic dependency generation
(-MMD), so we no longer need to explicitly list dependencies for each
source file, many of which were outdated anyway.

Also, most compilation is now done through generic pattern rules
per-directory, and only a few files still need their own rules.

And to make things more pretty, the lengthy compiler output is now
supressed by default, instead just printing "[CC] sourcefile.cpp". This
can be toggled by building with "make V=1".

ykempf committed:
Protect \\rho backslash.

ykempf committed:
Fix divideIfNonZero for negative denominators and include a threshold to guide in case of stray small values.

ykempf committed:
Protect the whistler speed divisions in electric field solver.

ursg committed:
Merge pull request #725 from ykempf/openmp_loops_in_grid_setup

Added two openmp for loops in LB weight assignment.

ursg committed:
Merge pull request #730 from ursg/makefileBeautification

Greatly simplify Makefile

ykempf committed:
Revert to differences in-code, but have actual derivatives in DROs.

ykempf committed:
Replace divideIfNonZero with divideIfNonNegative in cpu_moments.cpp.

ursg committed:
Print both relative and abso error variables in testpackage CI output

ursg committed:
Merge pull request #731 from ursg/CIrelativeAbsolute

Print both relative and abso error variables in testpackage CI output

ursg committed:
Makefile for Perlmutter (LBL).

Based on the LUMI makefile, as Perlmutter is the same architecture.
Basically only the library paths needed to be adjusted.

ykempf committed:
Revert "Fix divideIfNonZero for negative denominators and include a threshold to guide in case of stray small values."

This reverts commit 92cba4dd85854f53f6a909018a063aa57a3fb3ca.

ykempf committed:
Revert "Protect the whistler speed divisions in electric field solver."

This reverts commit aed73bda88c4b24e43023304909e29eedd4ba012.

ykempf committed:
Revert "Replace divideIfNonZero with divideIfNonNegative in cpu_moments.cpp."

This reverts commit 7aa43f510214dbe297f4abca3bc0dd254599b354.

ykempf committed:
Fix the revert in ldz_electric_field.cpp to compile.

ykempf committed:
Revert white-space changes to reduce clutter of diffs.

ursg committed:
Squelch one compilation warning (signed/unsigned comparison)

ursg committed:
Squelch warning about multi-line comment in ionosphere.

This used to be disabled by a command-line option, but the makefile
overhaul re-exposed it.

ursg committed:
Merge pull request #733 from ursg/oneWarningAway

Squelch one compilation warning (signed/unsigned comparison)

ursg committed:
Merge pull request #734 from ursg/squelchCommentWarning

Squelch warning about multi-line comment in ionosphere.

ursg committed:
Remove outdated warning about hardcoded meshName in ioread.

While technically correct, we are anyway hardcoding the other mesh names
in the corresponding fsgrid and ionosphere loading routines, too.
That warning basically had only historical value anymore.

ursg committed:
Replace deprecated inheritance of iterator types by type traits.

Newer compilers were complaining that OpenBucketHashtable's iterator
definitions were deprecated, because we were inheriting from
std::iterator.
I followed the guideline of
https://www.fluentcpp.com/2018/05/08/std-iterator-deprecated/ and
replaced them with type traits instead.

ursg committed:
Squelch warning about uninitialized neighborhood variable.

In faceNeighbor calculation, we always trust that the dimension values
we get back from DCCRG are sane.
This also introduces a safety check, which will hopefully never trigger.

ursg committed:
Merge pull request #735 from ursg/squelchMeshNameWarning

Remove outdated warning about hardcoded meshName in ioread.

ursg committed:
Merge pull request #736 from ursg/openBucketIteratorTypeTraits

Replace deprecated inheritance of iterator types by type traits.

ursg committed:
Remove warning for unused variable in vspace memory size calculation.

ursg committed:
Merge pull request #737 from ursg/squelchUninitializedVariableWarning

Squelch warning about uninitialized neighborhood variable.

ursg committed:
Squelch warning about unused IMFdelAx in vectordipole.cpp

ursg committed:
Squelch warning about IonosphereNodeInt datareducer diagnostic type.

The virtual function override wasn't working, because it expected an
int* instead of a Real*. Which is fundamentally ok, because it was
anyway replacing a no-op function by a no-op function.

But it's better to have the *correct* no-op doing nothing there.

Also, I like to reduce warnings.

ursg committed:
Merge pull request #738 from ursg/squelchUnusedVariableVBLOCKSIZE

Remove warning for unused variable in vspace memory size calculation.

ursg committed:
Merge pull request #739 from ursg/squelchOverloadHide

Squelch warning about IonosphereNodeInt datareducer diagnostic type.

ykempf committed:
Switching the order of early-exit phiprof calls in iowrite, the fix for TAU a month ago did them the wrong way around.

ykempf committed:
Merge pull request #741 from ykempf/fix_order_of_phiprof_calls

Switching the order of early-exit phiprof calls in iowrite, the fix f…

ursg committed:
Merge pull request #740 from ursg/squelchUnusedVariableIMFdelAx

Squelch warning about unused IMFdelAx in vectordipole.cpp

ursg committed:
Merge pull request #732 from ursg/perlmutter_makefile

Makefile for Perlmutter (LBL).

ursg committed:
Remove yet another unused variable varning in vspace memory calc.

This is basically the same fix as #738, in a slightly different
function.

ursg committed:
Remove warnings about some some unused "gridSize" variables in DROs.

ursg committed:
Merge pull request #743 from ursg/squelchUnusedVariableVBLOCKSIZE2

Remove yet another unused variable varning in vspace memory calc.

markusbattarbee committed:
Updated profiling in fieldsolver, corrected samples/Magnetosphere3d inflow magnetic field, made lumi makefile system library includes use -isystem

markusbattarbee committed:
moved timer creation inside parallel regions

markusbattarbee committed:
Changed timer type within omp parallel regions

markusbattarbee committed:
Changed timers also in derivatives, changed a bunch of collapse(3) loops to collapse(2)

ykempf committed:
Attempt at threading background field init loops.

ykempf committed:
Remove critical regions.

ursg committed:
Merge pull request #745 from ursg/squelchUnusedVariableDROs

Remove warnings about some some unused "gridSize" variables in DROs.

ursg committed:
Squelch a bunch of unused variable warnings.

ursg committed:
Merge pull request #747 from ursg/squelchUnusedVariables

Squelch a bunch of unused variable warnings.

ursg committed:
Squelch more unused variable warnings.

ykempf committed:
Thread the other function in backgroundfield, update comments.

ykempf committed:
Same threading fixes to the second loop.

ykempf committed:
Port *all* fixes.

ursg committed:
Remove some more unused variable and a signed/unsigned warning

ursg committed:
CI: Don't upload testpackage_build.log anymore, get proper return val.

Before, the pipe catching the compilation output was causing the job to
always return a green checkmark, even if the build failed for some
reason.

The information in that build log is anyway captured as part of the
github action, so there's no need to keep it around as an artefact.

ursg committed:
Merge pull request #750 from ursg/ci_compilation_status_fix

CI: Don't upload testpackage_build.log anymore, get proper return val.

ursg committed:
Squelch some more unused variable warnings.

ursg committed:
Merge pull request #749 from ursg/squelchEvenMoreWarnings

Remove some more unused variable and a signed/unsigned warning

ursg committed:
Merge pull request #752 from ursg/squelchYetnMoreWarnings

Squelch some more unused variable warnings.

ursg committed:
Squelch more unused variable warnings.

It's surprising how much little things have accumulated here and there,
that clearly served a purpose at some point in the past, but don't
anymore.

ursg committed:
Squelch even more unused variable warnings.

ykempf committed:
Merge branch 'dev' of github.com:fmihpc/vlasiator into dev

ursg committed:
Remove "#pragma ivdep" warnings by replacing with "#pragma omp simd"

ursg committed:
Merge pull request #753 from ursg/squelchJustABitMoreWarnings

Squelch some more unused variable warnings.

ursg committed:
Merge pull request #754 from ursg/squelchABitMoreWarnings

Squelch even more unused variable warnings.

ursg committed:
Fix three more unused variable warnings

ursg committed:
Merge pull request #756 from ursg/anotherSetOfWarnings

Fix three more unused variable warnings

ursg committed:
Squelch two unused variable warnings when not in debug mode.

(By moving the variables inside the same #define regions, where they
 would be used)

ursg committed:
Squelch a bunch of unused variables in ioread and iowrite

ursg committed:
Merge pull request #757 from ursg/squelchNDEBUGWarnings

Squelch two unused variable warnings when not in debug mode.

ursg committed:
Remove (you guessed it) some more unused variable warnings.

If you weigh a harddisk containing the code, it has probably become
lighter by a measurable amount already!

ursg committed:
Merge pull request #758 from ursg/IOReadWriteWarningsRemoval

Squelch a bunch of unused variables in ioread and iowrite

ursg committed:
Remove another unused variable, squelching one warning.

Turns out we never actually needed "end" in the backgroundfield
integration!

ursg committed:
Merge pull request #759 from ursg/willWeEverRunOutOfUnusedVariables

Remove (you guessed it) some more unused variable warnings.

henry2004y committed:
Add nonMaxwellianity DRO

ursg committed:
Squelch 3 warnings about uninitialized variables (in unused code).

This only actually warns if building with CLANG. Maybe GCC prunes the
unused code early enough to not even bother checking it?

ursg committed:
Merge pull request #760 from ursg/removeBGfieldEndWarning

Remove another unused variable, squelching one warning.

ursg committed:
Merge pull request #763 from ursg/uninitializedIChild

Squelch 3 warnings about uninitialized variables (in unused code).

ursg committed:
Remove unneeded include of boost/lexical_cast in readparameters.

This must have been some leftover from the past. Anyway, it's good to
reduce unneccessary dependencies.
(I actually found this when I tried to compile on my phone, it broke on
 all boost includes and I removed them one-by-one, until I noticed that
 I actually simply hadn't installed any boost package yet)

ursg committed:
Squelch two unused variable warnings in testAMR project.

The "myCell" variable is not being used when not doing debugging output,
so why not put it into the debugging #define as well!

ursg committed:
Merge pull request #766 from ursg/warningsReduceTestAMR

Squelch two unused variable warnings in testAMR project.

ursg committed:
Merge pull request #761 from henry2004y/nonMaxwellianity

Add nonMaxwellianity DRO

ursg committed:
Merge pull request #765 from ursg/removeBoostLexicalCast

Remove unneeded include of boost/lexical_cast in readparameters.

ursg committed:
Remove unused "olderr" variable from ionosphere.

This was a leftover from our experiments in solver schedule.

ursg committed:
Merge pull request #768 from ursg/warnigsReduceIonosphereOldErr

Remove unused "olderr" variable from ionosphere.

ursg committed:
Remove two unused functions from derivative.cpp

(They were trivial anyway, and thus probably never been taken into use)

ursg committed:
Remove some unused variables from gridGlue.

ursg committed:
Remove some unused variables in translation.

They probably were good for something at some point, but aren't anymore.

ursg committed:
Remove an unused variable from cpu_acc_map

ursg committed:
Merge pull request #771 from ursg/translationUnusedVariables

Remove some unused variables in translation.

ursg committed:
Merge pull request #772 from ursg/accelerationUnusedVariable

Remove an unused variable from cpu_acc_map

ursg committed:
Remove two unused variables from vlasovmover.cpp

(One is commented out, because some other commented-out timing code
 might depend on it)

lkotipal committed:
Fix orphaned timer

ursg committed:
Update PAPI github url in CI build script.

ursg committed:
Update PAPI github url in CI build script.

ursg committed:
Merge pull request #777 from ursg/fixCIPAPI

Update PAPI github url in CI build script.

ykempf committed:
Remove projections from ionosphere and conductingsphere perturbed B
computations.

lkotipal committed:
Make forced refinement lighter,
fix memory estimate for unrefine

ursg committed:
Merge pull request #774 from ursg/vlasovMoverUnusedVariables

Remove two unused variables from vlasovmover.cpp

ykempf committed:
Merge branch 'dev' of github.com:fmihpc/vlasiator into dev

ursg committed:
Fix (potentially) uninitialized return value warning in setbyuser.

Found by clang's warnings output.

ursg committed:
Merge pull request #780 from ursg/fixUninitializedSetbyUser

Fix (potentially) uninitialized return value warning in setbyuser.

kstppd committed:
Guard against nullptrs in shrink to fit

kstppd committed:
Removed extra line

kstppd committed:
Bailout when fg cells are not cubic

ursg committed:
Merge remote-tracking branch 'ykempf/threading_backgroundfield_init' into dev

markusbattarbee committed:
Added new cray programming environment makefile for Lumi

markusbattarbee committed:
renamed makefile to cpegnu

ykempf committed:
Merge pull request #781 from kstppd/shrink_to_fit_fix

Shrink to fit fix, fixes lots of segmentation faults.

ykempf committed:
Removal of the magnetic field projection function from system
boundaries.

ykempf committed:
Renaming conductingsphere to copysphere.

ykempf committed:
Last perfect conductor lines corrected.

ykempf committed:
Merge remote-tracking branch 'ykempf/make_differences_derivatives' into write_out_full_bgb_data_X_make_differences_derivatives

ykempf committed:
Output background bgb derivatives as derivatives too.

ykempf committed:
Write separate logfile for full BGB IO so as to not overwrite an
existing run's logfile.txt.

ykempf committed:
Removed writing of background derivatives with the fg_derivs DRO.
Cleaned up metadata too.

ykempf committed:
More metadata inconsistency fixes.

ykempf committed:
More consistency and metadata details.

ykempf committed:
Add flags to write all DROs to file or diagnostic.
This is influenced by the flag to write or not fg variables.

lkotipal committed:
Revert "Update path to dccrg"

This reverts commit 0d6ff9409ba2f6aed217f246d750eab60ed7bf0c.

lkotipal committed:
Revert "Dummy commit"

This reverts commit 82ea123cd8ecdd8bb445379b5f14fd629cbd7159.

ykempf committed:
Merge pull request #788 from markusbattarbee/lumi_cpe_makefile

LUMI-C PrgEnv-gnu Makefile

lkotipal committed:
Merge branch 'dev' into cache-neighbors-dummy

lkotipal committed:
Structured binding

lkotipal committed:
Merge branch 'dev' into timer-object

markusbattarbee committed:
Merge pull request #717 from lkotipal/cache-neighbors-dummy

Cache neighbors

ykempf committed:
Added some \\Deltas to the metadata.

ursg committed:
Remove commented-out stencilWidth from gridGlue

ykempf committed:
Set io.system_write_all_data_reducers and
io.diagnostic_write_all_data_reducers in testapakcage Ionosphere_small
test.

ykempf committed:
Merge pull request #783 from kstppd/bailout_non_cubic_fg_cells

Bailout when fg cells are not cubic

ykempf committed:
Remove useless assignment at end of function.

ykempf committed:
Merge pull request #770 from ursg/gridGlueUnusedVariables

Remove some unused variables from gridGlue.

ykempf committed:
Merge branch 'dev' into write_out_full_bgb_data_X_make_differences_derivatives

ykempf committed:
Fix merge blunder.

ykempf committed:
Merge pull request #789 from ykempf/remove_acc_unneeded_assignment

Remove useless assignment at end of function.

ykempf committed:
Merge pull request #776 from lkotipal/timer-fix

Fix orphaned timer

ykempf committed:
Comments to explain better the full BGB writing procedure.

ykempf committed:
Setup ionosphere communicator for full BGB output including ionosphere
grid metadata.

ykempf committed:
Override output variable list when writing full BGB.

ykempf committed:
Whitespace fix.

markusbattarbee committed:
More reorderings. Whitespace changes due to M-x-delete-trailing-whitespace.

markusbattarbee committed:
Merge branch 'dev' into profile_fields

ykempf committed:
Fix order of phiprof calls upon early exit.

ykempf committed:
Comments and fixes in the derivatives IO.

ykempf committed:
Fix volume background field derivatives integration loop to get dbidi and remove the fsgrid
background field dbidi derivatives.

ykempf committed:
Derivative data reducers to subfolders.

ykempf committed:
Merge pull request #784 from ykempf/inner_boundary_plain_copy

Remove projections from ionosphere and conductingsphere perturbed B computations

ykempf committed:
Shorten the max distance for full box tracing.

ykempf committed:
Merge branch 'dev' of github.com:fmihpc/vlasiator into dev

ursg committed:
Makefile and a small fix to build on jetsonTX Arm cluster at BSC.

Note that the type change in ionosphere.cpp is also required to build on
RISC-V, so I suppose this is relevant for all architectures with some
alignment constraints. Not a performance issue either, so why not just
fix it for all.

ykempf committed:
Include background constant magnetic field in calculateIonosphereFsgridCoupling recording of upmapped B.

ykempf committed:
Ghost update and derivative computation timing fixes for ionosphere coupling.

ykempf committed:
Remove hemisphere swap check, stop coupling tracing if hitting back ionosphere inner radius. No warning printed.

ykempf committed:
More accurate tracing to inner and coupling radii (copy of Markku's interpolation), and fix of MPI domain edge bug.

lkotipal committed:
LUMI makefile fixes

lkotipal committed:
As per documentation

ykempf committed:
Update to my own local Makefile.

ykempf committed:
MPI_T used to force disabling OpenMPI 4 ompiio (no checks for other
pre-set flags, this has highest precedence, so probably more
delicate-handed approach incoming before merging).

ykempf committed:
Detect MPI library and disable ompio for OpenMPI.

ykempf committed:
Makefile flag to disable MCA io override in code.

ykempf committed:
Some formatting and simplification.

kstppd committed:
Just abort when cells are not cubic and do not use the bailout mechanism which ends up writing a restart

kstppd committed:
Upate message

markusbattarbee committed:
Merge pull request #801 from kstppd/non-uniform-bailout

Abort when cells are not cubic.

ykempf committed:
Recompute full conductivity tensor at restart to fix ig_inplanecurrent
output occurring between retsrat and next ionosphere solution.

ykempf committed:
Batch change of ionosphere variable descriptions to doxygen style.

ykempf committed:
Remove debugging cerr

ykempf committed:
Make fullbox and fluxrope tracing max distance an optional cfg parameter.

ykempf committed:
Fixed order of MPI_Finalize and MPI_T_finalize.

lkotipal committed:
Merge branch 'dev' into timer-object

lkotipal committed:
Convert more timers

lkotipal committed:
Revert "Use updated phiprof (do not merge)"

This reverts commit 415d8a161a2ba563ad0711fb26817970c2ae607b.

lkotipal committed:
Don't purge

lkotipal committed:
Early return

markusbattarbee committed:
Merge pull request #787 from ykempf/write_all_dros

Write all DROs

markusbattarbee committed:
Merge pull request #769 from ursg/removeTwoUnusedFunctions

Remove two unused functions from derivatives.cpp

markusbattarbee committed:
Merge branch 'dev' into vscode-support

markusbattarbee committed:
Merge pull request #723 from lkotipal/vscode-support

Visual Studio Code support

markusbattarbee committed:
Merge branch 'dev' into profile_fields

markusbattarbee committed:
Merge branch 'dev' into profile_fields

markusbattarbee committed:
Removed dynamic(1) from ldz magnetic field loop pragma

markusbattarbee committed:
Merge branch 'profile_fields' of github.com:markusbattarbee/vlasiator-public into profile_fields

ursg committed:
CI: error if artefact upload fails

markusbattarbee committed:
Merge pull request #822 from ursg/artefactError

CI: error if artefact upload fails

markusbattarbee committed:
Added sleep command to testpackage creation in CI so upload might work

lkotipal committed:
Do not clip alpha

lkotipal committed:
Revert "Remove two unused functions from derivative.cpp"

This reverts commit 543bba30fe05563a92d5c1b82c931d9a84a9e282.

lkotipal committed:
Remove unused function

ursg committed:
Merge remote-tracking branch 'ykempf/write_out_full_bgb_data_X_make_differences_derivatives' into HEAD

ursg committed:
Merge pull request #793 from ykempf/shorten_max_tracing_distance

Shorten the max distance for full box tracing.

ursg committed:
Merge pull request #798 from lkotipal/new-makefiles

LUMI makefile fixes

ursg committed:
Merge pull request #795 from ursg/jetsonTX

Makefile and a small fix to build on jetsonTX Arm cluster at BSC.

ykempf committed:
Trailing white space (trigger CI)

ursg committed:
Merge pull request #721 from lkotipal/timer-object

Timer object

ursg committed:
Merge pull request #796 from ykempf/fieldtracing_fixes

Fieldtracing fixes

ursg committed:
Merge branch 'dev' into MPI_T_MCA_disabling_mpiio

lkotipal committed:
Merge branch 'dev' into amr-dev

lkotipal committed:
Remove leastCellsToRefine

ursg committed:
Merge pull request #755 from ursg/removeIvdepWarnings

Remove "#pragma ivdep" warnings by replacing with "#pragma omp simd"

markusbattarbee committed:
Brought fieldsolver profiling up to date with dev and new timer usage

ykempf committed:
Merge branch 'dev' into write_out_full_bgb_data_X_make_differences_derivatives

lkotipal committed:
Documentation

lkotipal committed:
Timer fixes

ykempf committed:
Two continue not checking against writing all DROs now fixed.

markusbattarbee committed:
Removed some comments to trigger new CI

ykempf committed:
Merge pull request #786 from ykempf/write_out_full_bgb_data_X_make_differences_derivatives

Write out full bgb data x make differences derivatives

ykempf committed:
Remove ompio flags in scripts and Makefile comments.

markusbattarbee committed:
Merge branch 'dev' into profile_fields

ykempf committed:
Merge branch 'dev' into ig_inplanecurrent_at_restart

ykempf committed:
sleep 10s before ending testpackage building so uploading gets a better chance on carrington

ykempf committed:
Merge pull request #744 from markusbattarbee/profile_fields

Updated profiling in fieldsolver, corrected samples/Magnetosphere3d i…

ykempf committed:
Merge pull request #802 from ykempf/ig_inplanecurrent_at_restart

ig_inplanecurrent at restart

ykempf committed:
Merge branch 'dev' into wait_before_upload

ykempf committed:
Merge pull request #827 from ykempf/wait_before_upload

sleep 10s before ending testpackage building so uploading gets a better chance on carrington
I know it's double use but the one merged from Markus still failed a couple times.

ykempf committed:
Merge pull request #800 from ykempf/MPI_T_MCA_disabling_mpiio

MPI_T MCA disabling mpiio

ykempf committed:
Merge branch 'dev' into amr-dev

alhom committed:
Add some checks to make sure we actually get all the neighborhoods we want to get, abort if not.

alhom committed:
Fix an output

ursg committed:
Merge pull request #834 from alhom/add_neighborhood_checks

Add neighborhood checks

markusbattarbee committed:
Also output current library commits to version.cpp

markusbattarbee committed:
Updated mahti makefile, better handling of non-existing git repositories in commit version output.

ykempf committed:
Merge pull request #836 from markusbattarbee/print_library_commits_to_version

Also output current library commits to version.cpp

ykempf committed:
Update to Makefile.yann

ykempf committed:
Merge pull request #838 from ykempf/update_makefile

Update to Makefile.yann

lkotipal committed:
More descriptive names

ursg committed:
Merge pull request #701 from lkotipal/amr-dev

AMR Update

ursg committed:
Try to squelch CI ucx warnings.

(These were spamming the log files, thus exceeding the 64k check
 annotation limit)

ursg committed:
Fix whitespace typo.

ursg committed:
Merge pull request #840 from ursg/CI_ucx_errors

Try to squelch CI ucx warnings.

ykempf committed:
Merge remote-tracking branch 'origin/dev' into fang_ionospheric_precipitation

ykempf committed:
Fixes to merge from dev.

ykempf committed:
Americanization

ykempf committed:
Update BVOL on mpiGrid before using it in proton precipitation.

ykempf committed:
Stray 1e3 removed.

ykempf committed:
Comment on update of BVOL.

ykempf committed:
Filter negative values in proton spectrum computation.

ykempf committed:
Divide by solid angle to compute precipitating proton flux.
